python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,227 @@
+            req: the JSON formatted request that contains two string fields: 'event_key' and 'event_payload'
+                'event_key' uniquely identifies a node in the receiving workflow;
+                'event_payload' refers to the event's content
+        Example:
+            JSON formatted request {"event_key":"node_event","event_payload":"approved"}
+        Returns:
+            if the event was received and processed, HTTP response status 200
+            if the event was not expected or the workflow_id did not exist, HTTP response status 404
+            if the event was received but failed at checkpointing, HTTP response 500
+
+        """
+        req_json = await req.json()
+        try:
+            event_key = req_json['event_key']
+        except KeyError:
+            return JSONResponse(
+                status_code=404,
+                content={"error": {"code": 404,
+                                   "message": "event_key field is not found in the request JSON"}}
+            )
+        try:
+            event_payload = req_json['event_payload']
+        except KeyError:
+            return JSONResponse(
+                status_code=404,
+                content={"error":
+                             {"code": 404,
+                                   "message": "event_payload field is not found in the request JSON"}}
+            )
+        try:
+            self.event_key_payload[workflow_id][event_key].set_result((event_key, event_payload))
+        except KeyError:
+            return JSONResponse(
+                status_code=404,
+                content={"error":
+                             {"code": 404,
+                                   "message": "workflow_id or event_key is not registered to receive event"}}
+            )
+
+        self.event_checkpoint_pending[event_key] = asyncio.Future()
+        confirmed = await self.event_checkpoint_pending[event_key]
+        self.event_checkpoint_pending.pop(event_key)
+        if confirmed:
+            return JSONResponse(
+                status_code=200,
+                content={}
+            )
+        return JSONResponse(
+            status_code=500,
+            content={"error": {"code": 500,
+                               "message": "event processing failed"}}
+        )
+
+    async def get_event_payload(self, workflow_id:str, event_key:str) -> Event:
+        """Internal method used by HTTPListner to subscribe to an event matched by workflow_id and event_key
+        """
+        try:
+            self.event_key_payload[workflow_id]
+        except KeyError:
+            self.event_key_payload[workflow_id] = {}
+        else:
+            try:
+                self.event_key_payload[workflow_id][event_key]
+            except KeyError:
+                # expected since event_key should not exist
+                pass
+            else:
+                raise WorkflowEventHandleError(workflow_id, f"The same {event_key} is used to get payload again.")
```

Comment:
The pattern is wired...

You should just do:

```
if workflow_id not in self.event_key_payload:
    self.event_key_payload[workflow_id] = {}

if event_key in self.event_key_payload[workflow_id]:
    raise WorkflowEventHandleError(workflow_id, f"The same {event_key} is used to get payload again.")
```

Reply:
Done